synapse/storage/databases/main/events_worker.py

```diff
@@ -1626,3 +1626,173 @@ def _cleanup_old_transaction_ids_txn(txn):
+            # If the event in question has its prev_events listed as a backward
+            # extremity, it's next to a gap.
+            #
+            # We can't just check the backward edges in `event_edges` because
+            # when we persist event, we will also record the prev_events as
+            # edges to the event in question regardless of if we have those
+            # prev_events yet. We need to check whether those prev_events are
+            # backward extremity also known as gaps that need to be backfilled.
+            backward_extremity_query = """
+                SELECT 1 FROM event_backward_extremities
+                WHERE
+                    room_id = ?
+                    AND event_id = ?
+                LIMIT 1
+            """
+
+            # If the event in question is a forward extremity, we will just
+            # consider any potential forward gap as not a gap since it's one of
+            # the latest events in the room.
+            #
+            # We can't combine this query with the `forward_edge_query` below
+            # because if the event in question has no forward edges(isn't
+            # referenced by any other event's prev_events) but is in
+            # `event_forward_extremities`, we don't want to return 0 rows and
+            # say it's next to a gap.
+            forward_extremity_query = """
+                SELECT 1 FROM event_forward_extremities
+                WHERE
+                    room_id = ?
+                    AND event_id = ?
+                LIMIT 1
+            """
+
+            # Check to see whether the event in question is already referenced
+            # by another event. If we don't see any edges, we're next to a
+            # forward gap.
+            #
+            # We can't just only check `event_forward_extremities` directly
+            # because that doesn't include backfilled and outlier events.
+            forward_edge_query = """
+                SELECT 1 FROM event_edges
+                /* Check to make sure the event referencing our event in question is not rejected */
+                LEFT JOIN rejections ON event_edges.event_id == rejections.event_id
+                WHERE
+                    event_edges.room_id = ?
+                    AND event_edges.prev_event_id = ?
+                    /* It's not a valid edge if the event referencing our event in
+                     * question is rejected.
+                     */
+                    AND rejections.event_id IS NULL
+                LIMIT 1
+            """
+
+            # If the event in question has its prev_events listed as a backward
+            # extremity, it's next to a backwards gap.
+            #
+            # We need to check this before the forward edges below as the event
+            # in question could be a forward extremity while still having a
+            # backwards gap.
+            for prev_event_id in event.prev_event_ids():
```

Comment:
Calling out a bigger change here since the last review cycle 👆

Previously, there was a logic flaw in the way I was detecting backwards gaps where I was just trying to check if the event in question was a backward extremity. But this would never be the case because when we persist an event, we clear it as a backward extremity and record it's `prev_events` as the new backward extremities.

So we really need to just whether the event in question has it's `prev_events` listed as backward extremity which is what the code is doing now.

Reply:
I'm confused about what a backward extremity is now. Previously I thought it was an event we had, but whose `prev_events` we did not all have. After looking at the [code](https://github.com/matrix-org/synapse/blob/a4521ce0a8d252e77ca8bd261ecf40ba67511a31/synapse/storage/databases/main/events.py#L2136), it seems to agree with what you're saying.

When I prod at my synapse's database, I see a lot of events in `event_backward_extremities` that are also in `events` and some that aren't. I'm guessing the ones that are in `events` are outliers. Does an outlier always have one `prev_event` that is also a backward extremity?

Reply:
>  Does an `outlier` always have one `prev_event` that is also a backward extremity?

Not necessarily:

 1. I think an `outlier` could reference a a non-outlier event that we already have persisted. Imagine if we backfill an event from `bob` and we're missing a `prev_event`. Synapse will go and fetch the missing `prev_event` and store it as an `outlier` but that event could be referencing events we already have persisted normally.
    [Mermaid live editor playground](https://mermaid-js.github.io/mermaid-live-editor/edit/#eyJjb2RlIjoiZmxvd2NoYXJ0IEJUXG4gICAgc3ViZ3JhcGggcGVyc2lzdGVkXG4gICAgICAgIEMtLS0tPkItLS0-QVxuICAgIGVuZFxuXG4gICAgc3ViZ3JhcGggYmFja2ZpbGxlZFxuICAgICAgICBib2JFdmVudFxuICAgIGVuZFxuXG4gICAgYm9iRXZlbnQtLT5tZW1iZXJCb2IwKFtcIm0ucm9vbS5tZW1iZXIgKGJvYilcIl0pXG5cbiAgICBtZW1iZXJCb2IwLS0-QiIsIm1lcm1haWQiOiJ7XG4gIFwidGhlbWVcIjogXCJkZWZhdWx0XCJcbn0iLCJ1cGRhdGVFZGl0b3IiOmZhbHNlLCJhdXRvU3luYyI6dHJ1ZSwidXBkYXRlRGlhZ3JhbSI6ZmFsc2V9)
    ![](https://user-images.githubusercontent.com/558581/144124178-f4af9082-ced4-4d15-8053-a79b3cbf5e01.png)
 1. Events can be created with no `prev_events`. Related discussion where I am playing around no `prev_event` events for MSC2716, https://github.com/matrix-org/synapse/pull/11243#discussion_r750816377

---

> I'm confused about what a backward extremity is now.

Looking at the code, [`outliers` are never cleared as a `event_backward_extremities`](https://github.com/matrix-org/synapse/blob/70cbb1a5e311f609b624e3fae1a1712db639c51e/synapse/storage/databases/main/events.py#L2214). So the `event_backward_extremities` you also see persisted in `events` are `outliers`. 

The other backward extremities are just the oldest events we know about, not necessarily persisted. Backward extremities just need to be `event_id` strings and will most likely be the `prev_events` of the oldest events we have persisted.

I wrote a bit about them in https://github.com/matrix-org/synapse/blob/develop/docs/development/room-dag-concepts.md#backwards-extremity (also see the `outliers` section) but I think it needs another iteration to get it more correct.

Reply:
Thanks for clearing things up for me!